File: openadmin/spec/pie_chart.py
# SPDX-FileCopyrightText: 2026 OpenAdmin
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from typing import Literal, NotRequired, TypedDict

from .colors import Color
from .error import Error
from .http_methods import HttpMethod
from .icons import Icon
from .json_schema import JsonSchema


class PieChartConfigValue(TypedDict):
    name: NotRequired[str]
    color: NotRequired[Color]
    icon: NotRequired[Icon]


class PieChartComponent(TypedDict):
    type: Literal["pie-chart"]
    id: str
    name: str
    config: dict[str, PieChartConfigValue] | None
    icon: Icon | None
    name_key: str | None
    value_key: str | None
    color: Color | None
    description: str | None
    caption: str | None
    caption_description: str | None
    caption_icon: Icon | None
    method: HttpMethod
    form: JsonSchema | None
    body: JsonSchema | None
    query: JsonSchema | None


type PieChartData = (
    list[dict[str, int | float | str]]
    | list[dict[Literal["name", "value"], int | float | str]]
    | object
)


class PieChartResponce(TypedDict):
    config: NotRequired[dict[str, PieChartConfigValue]]
    icon: NotRequired[Icon]
    color: NotRequired[Color]
    data: PieChartData


type PieChart = PieChartData | PieChartResponce | Error

File: client/src/schemas/pie-chart.ts
// SPDX-FileCopyrightText: 2026 OpenAdmin
//
// SPDX-License-Identifier: AGPL-3.0-or-later

import { z } from "zod"
import { colorSchema } from "./color"
import { httpMethodSchema } from "./http-method"
import { iconSchema } from "./icon"
import { jsonSchemaSchema } from "./json-schema"

export const pieChartConfigValueSchema = z.object({
	name: z.string().optional(),
	color: colorSchema.optional(),
	icon: iconSchema.optional(),
})

export type PieChartConfigValue = z.infer<typeof pieChartConfigValueSchema>

export const pieChartComponentSchema = z.object({
	type: z.literal("pie-chart"),
	id: z.string(),
	name: z.string(),
	config: z.record(z.string(), pieChartConfigValueSchema).nullable(),
	icon: iconSchema.nullable(),
	name_key: z.string().nullable(),
	value_key: z.string().nullable(),
	color: colorSchema.nullable(),
	description: z.string().nullable(),
	caption: z.string().nullable(),
	caption_description: z.string().nullable(),
	caption_icon: iconSchema.nullable(),
	method: httpMethodSchema,
	form: jsonSchemaSchema.nullable(),
	body: jsonSchemaSchema.nullable(),
	query: jsonSchemaSchema.nullable(),
})

export type PieChartComponent = z.infer<typeof pieChartComponentSchema>

export const pieChartDataSchema = z.union([
	z.array(z.record(z.string(), z.union([z.number(), z.string()]))),
	z.array(z.record(z.enum(["name", "value"]), z.union([z.number(), z.string()]))),
	z.unknown(),
])

export type PieChartData = z.infer<typeof pieChartDataSchema>

export const pieChartResponseSchema = z.object({
	config: z.record(z.string(), pieChartConfigValueSchema).optional(),
	icon: iconSchema.optional(),
	color: colorSchema.optional(),
	data: pieChartDataSchema,
})

export type PieChartResponse = z.infer<typeof pieChartResponseSchema>

export const pieChartSchema = z.union([pieChartDataSchema, pieChartResponseSchema])

export type PieChart = z.infer<typeof pieChartSchema>
